src/coreclr/nativeaot/Runtime/DebugHeader.cpp

```diff
@@ -0,0 +1,275 @@
+    MAKE_DEBUG_FIELD_ENTRY(GcDacVars, minor_version_number);
+    MAKE_DEBUG_FIELD_ENTRY(GcDacVars, generation_size);
+    MAKE_DEBUG_FIELD_ENTRY(GcDacVars, total_generation_count);
+    MAKE_DEBUG_FIELD_ENTRY(GcDacVars, built_with_svr);
+    MAKE_DEBUG_FIELD_ENTRY(GcDacVars, finalize_queue);
+    MAKE_DEBUG_FIELD_ENTRY(GcDacVars, generation_table);
+    MAKE_DEBUG_FIELD_ENTRY(GcDacVars, ephemeral_heap_segment);
+    MAKE_DEBUG_FIELD_ENTRY(GcDacVars, alloc_allocated);
+    MAKE_DEBUG_FIELD_ENTRY(GcDacVars, n_heaps);
+    MAKE_DEBUG_FIELD_ENTRY(GcDacVars, g_heaps);
+
+    MAKE_SIZE_ENTRY(dac_gc_heap);
+    MAKE_DEBUG_FIELD_ENTRY(dac_gc_heap, alloc_allocated);
+    MAKE_DEBUG_FIELD_ENTRY(dac_gc_heap, ephemeral_heap_segment);
+    MAKE_DEBUG_FIELD_ENTRY(dac_gc_heap, finalize_queue);
+    MAKE_DEBUG_FIELD_ENTRY(dac_gc_heap, generation_table);
+
+    MAKE_SIZE_ENTRY(gc_alloc_context);
+    MAKE_DEBUG_FIELD_ENTRY(gc_alloc_context, alloc_ptr);
+    MAKE_DEBUG_FIELD_ENTRY(gc_alloc_context, alloc_limit);
+    MAKE_DEBUG_FIELD_ENTRY(gc_alloc_context, alloc_bytes);
+    MAKE_DEBUG_FIELD_ENTRY(gc_alloc_context, alloc_bytes_uoh);
+    MAKE_DEBUG_FIELD_ENTRY(gc_alloc_context, alloc_count);
+
+    MAKE_SIZE_ENTRY(dac_generation);
+    MAKE_DEBUG_FIELD_ENTRY(dac_generation, allocation_context);
+    MAKE_DEBUG_FIELD_ENTRY(dac_generation, start_segment);
+    MAKE_DEBUG_FIELD_ENTRY(dac_generation, allocation_start);
+
+    MAKE_SIZE_ENTRY(dac_heap_segment);
+    MAKE_DEBUG_FIELD_ENTRY(dac_heap_segment, allocated);
+    MAKE_DEBUG_FIELD_ENTRY(dac_heap_segment, committed);
+    MAKE_DEBUG_FIELD_ENTRY(dac_heap_segment, reserved);
+    MAKE_DEBUG_FIELD_ENTRY(dac_heap_segment, used);
+    MAKE_DEBUG_FIELD_ENTRY(dac_heap_segment, mem);
+    MAKE_DEBUG_FIELD_ENTRY(dac_heap_segment, flags);
+    MAKE_DEBUG_FIELD_ENTRY(dac_heap_segment, next);
+    MAKE_DEBUG_FIELD_ENTRY(dac_heap_segment, background_allocated);
+    MAKE_DEBUG_FIELD_ENTRY(dac_heap_segment, heap);
+
+    MAKE_DEFINE_ENTRY(FinalizeExtraSegCount, dac_finalize_queue::ExtraSegCount);
+    MAKE_DEFINE_ENTRY(MinObjectSize, MIN_OBJECT_SIZE);
+
+    MAKE_SIZE_ENTRY(ThreadStore);
+    MAKE_DEBUG_FIELD_ENTRY(ThreadStore, m_ThreadList);
+
+    MAKE_SIZE_ENTRY(ThreadBuffer);
+    MAKE_DEBUG_FIELD_ENTRY(ThreadBuffer, m_pNext);
+    MAKE_DEBUG_FIELD_ENTRY(ThreadBuffer, m_rgbAllocContextBuffer);
+    MAKE_DEBUG_FIELD_ENTRY(ThreadBuffer, m_threadId);
+    MAKE_DEBUG_FIELD_ENTRY(ThreadBuffer, m_pThreadStressLog);
+
+    // EEThreadID is forward declared and not available
+    MAKE_DEBUG_ENTRY(EEThreadID, SIZEOF, sizeof(void*));
+    MAKE_DEBUG_ENTRY(EEThreadID, m_FiberPtrId, 0);
+
+    MAKE_SIZE_ENTRY(EEType);
+    MAKE_DEBUG_FIELD_ENTRY(EEType, m_uBaseSize);
+    MAKE_DEBUG_FIELD_ENTRY(EEType, m_usComponentSize);
+    MAKE_DEBUG_FIELD_ENTRY(EEType, m_usFlags);    MAKE_DEBUG_ENTRY(EEType, m_pBaseType, offsetof(EEType, m_RelatedType) + offsetof(EEType::RelatedTypeUnion, m_pBaseType));
```

Comment:
It would be good to capture somewhere which flags the debugger is going to depend on. We have been changing these in the past.

Reply:
The flags are used for !verifyheap, the logic is copied from [EEType::Validate](https://github.com/dotnet/runtimelab/blob/99dbab06575ab15fac0fbfd53df1ecd863914fbe/src/coreclr/nativeaot/Runtime/eetype.cpp#L19) for object validation. 

I assumed that the flags were stable, I can add some static asserts though that check that they change and instruct to bump the debug header version if they are different.

Reply:
Asserts are added now